### PR TITLE
Add Test Plans

### LINKED ISSUE #684: Support for TestPlan
As starting with Xcode11 Apple introduces TestPlans, XcodeGen still doesn't have support to generate project using Test Plans.

### PR DESCRIPTION
**Pull request by yonaskolb**
Adds test plan support
Resolves #684 

- [x] reference test plans in schemes
- [x] generate test plan references in schemes
- [x] parsing tests
- [x] generation tests
- [x] updates scheme versions
- [x] remove xctestplan from resources
- [x] validation errors
- [x] validation tests
- [ ] ability to define test plans in spec (probably not required) OR:
- [ ] way to update UUID references in test plan files
- [x] docs

Possible changes:
- make top level `testPlans` which are then referenced in schemes, similar to the new `projectReferences`?
- ability to choose a default test plan? At the moment the first test plan in a scheme will be the default
- ability to define test plans in the spec itself? If not we probably need a way to update the UUIDs in existing files in case XcodeGen/XcodeProj ever changes them

### CODE REVIEW

File: Sources/ProjectSpec/Scheme.swift [hunk 1]
```diff
@@ -318,6 +332,15 @@ extension Scheme.Run: JSONEncodable {
     }
 }
 
+extension Scheme.Test: PathContainer {
+
+    static var pathProperties: [PathProperty] {
+        [
+            .string("testPlans")
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift]: This file is getting large, gonna split it up

File: Sources/XcodeGenKit/SchemeGenerator.swift [hunk 2]
```diff
@@ -194,10 +194,18 @@ public class SchemeGenerator {
         let launchVariables = scheme.run.flatMap { $0.environmentVariables.isEmpty ? nil : $0.environmentVariables }
         let profileVariables = scheme.profile.flatMap { $0.environmentVariables.isEmpty ? nil : $0.environmentVariables }
 
+        var testPlans: [XCScheme.TestPlanReference]?
```
Comment (giginet) [inline Sources/XcodeGenKit/SchemeGenerator.swift]: You can use `let`

```swift
let testPlanns: [XCScheme.TestPlanReference] = scheme.test?.testPlans.map { plans in
    return plans.enumerated().map { index, path in
        .init(reference: "container:\(path)", default: index == 0)
    }
}
```

Comment (yonaskolb) [reply Sources/XcodeGenKit/SchemeGenerator.swift]: Yeah there’s a few ways to do it. The property has to be nil if there are no test plans though, as that’s what XcodeProj expects

Comment (giginet) [reply Sources/XcodeGenKit/SchemeGenerator.swift]: Points of my comment was avoiding using mutable variables as we can. 
Type should be `[XCScheme.TestPlanReference]?` Sorry.

**Comment by min** [comment]
@yonaskolb any updates on release for this? would love to start migrating to test plans.

**Comment by armintelker** [comment]
I really like to use XcodeGen but our project is build up on test plans. Is here any update.

**Comment by soffes** [comment]
Really looking forward to this! Thanks for your effort on this feature! <3

**Comment by pinda** [comment]
Looking forward to this! Any ETA on when this will become available?

**Comment by omares** [comment]
hey @yonaskolb / @giginet is there anything one can do to help out finializing this? I would like to give it a try and pick this up.

**Comment by omares** [comment]
Went ahead and created an [up-to-date version of the branch](https://github.com/yonaskolb/XcodeGen/compare/master...omares:test_plans_continued?expand=1), tried my best to resolve the conflicts, but i really need some hints what the actual code was lacking. :D

**Comment by omares** [comment]
I created a draft PR #846 that contains the default test plan option, validation errors & tests and documentation. It only lacks the uuid update, would love to hear hints on how to tackle that. I would suggest taking further discussions, on how to proceed to the [new PR](https://github.com/yonaskolb/XcodeGen/pull/846).

**Comment by popwarfour** [comment]
Would love this feature!

**Comment by Almaz5200** [comment]
Hi! Any updates on when this PR is going to be merged?

**Comment by vCrespoP** [comment]
+1 to test plans!

**Comment by DmitryShapovalov** [comment]
also, +1 to test plans!

**Comment by SarannyaSalim** [comment]
+1 to test plans

**Comment by rublagar** [comment]
+1 to test plans

**Comment by hakkurishian** [comment]
+++++++1 :)

**Comment by yonaskolb** [comment]
Ok, have found some time to pick this up! I've pulled in changes from https://github.com/yonaskolb/XcodeGen/pull/846 and also made some further additions.
This is now at a point where it's usable with some big caveats for now. As the docs now say:

> For now test plans are not generated by XcodeGen and must be created in Xcode and checked in, and then referenced by path. If the test targets are added, removed or renamed, the test plans may need to be updated in Xcode.

I hope this is useable in its current state for some people. The plan is to release this in the next version.

In terms of fully specced and generated test plans that is a whole bunch more work:
- Support for these types of files in XcodeProj https://github.com/tuist/XcodeProj/issues/484
- The addition of Test Plans in the project spec. The test plan file format is also expected to be a configuration file that changes often with Xcode releases making the spec hard to keep up to date.

**Comment by yonaskolb** [comment]
Released in 2.29.0

**Pull request merged**